engine/userguide/storagedriver/btrfs-driver.md

```diff
@@ -227,179 +106,193 @@ loaded. To verify this, use the following command:
+
+### Writing files
+
+- **Writing new files**: Writing a new file to a container invokes an allocate-on-demand
+  operation to allocate new data block to the container's snapshot. The file is
+  then written to this new space. The allocate-on-demand operation is native to
+  all writes with Btrfs and is the same as writing new data to a subvolume. As a
+  result, writing new files to a container's snapshot operates at native Btrfs
+  speeds.
+
+- **Modifying existing files**: Updating an existing file in a container is a copy-on-write
+  operation (*redirect-on-write* is the Btrfs terminology). The original data is
+  read from the layer where the file currently exists, and only the modified
+  blocks are written into the container's writable layer. Next, the Btrfs driver\
+  updates the filesystem metadata in the snapshot to point to this new data.
+  This behavior incurs very little overhead.
+
+- **Deleting files or directories**: If a container deletes a file or directory
+  that exists in a lower layer, Btrfs masks the existence of the file or
+  directory in the lower layer. If a container creates a file and then deletes
+  it, this operation is performed in the Btrfs filesystem itself and the space
+  is reclaimed.
 
-    ```bash
-    $ sudo rm -rf /var/lib/docker.bk
-    ```
+With Btrfs, writing and updating lots of small files can result in slow
+performance.
 
 ## Btrfs and Docker performance
 
 There are several factors that influence Docker's performance under the `btrfs`
  storage driver.
 
 - **Page caching**. Btrfs does not support page cache sharing. This means that
-*n* containers accessing the same file require *n* copies to be cached. As a
-result, the `btrfs` driver may not be the best choice for PaaS and other high
-density container use cases.
-
-- **Small writes**. Containers performing lots of small writes (including
-Docker hosts that start and stop many containers) can lead to poor use of Btrfs
- chunks. This can ultimately lead to out-of-space conditions on your Docker
-host and stop working. This is currently a major drawback to use current
-versions of Btrfs.
+  each accessing the same file copies the file into the Docker hosts's memory.
+  As a result, the `btrfs` driver may not be the best choice high-density use
+  cases such as PaaS.
 
-    If you use the `btrfs` storage driver, closely monitor the free space on
-your Btrfs filesystem using the `btrfs filesys show` command. Do not trust the
-output of normal Unix commands such as `df`; always use the Btrfs native
-commands.
+- **Small writes**. Containers performing lots of small writes (this usage
+  pattern matches what happens when you start and stop many containers in a short
+  period of time, as well) can lead to poor use of Btrfs chunks. This can
+  prematurely fill the Btrfs filesystem and lead to out-of-space conditions on
+  your Docker host. Use `btrfs filesys show` to closely monitor the amount of
+  free space on your Btrfs device.
```

Comment:
Perhaps we should refer here to the use of volumes already?

Tangential, but most _write_ issues on our storage driver are not an issue if the containers are "well designed", and are considered immutable (only some writes); any high-traffic data should not live in the container's filesystem